ref(metrics): Base sessions v2 API on metrics query builder [INGEST-1193 INGEST-1229]

jjbayer opened the pull request:
Instead of talking to snuba and the metrics indexer directly, compose a query to the metrics API to fetch the data served from the sessions API. Beside removing code duplication, this has the advantage that new derived metrics (e.g. `crash_rate`) can be easily exposed through the sessions API.

jjbayer committed:
wip: Interface definition

jjbayer committed:
wip

jjbayer committed:
wip: First test passes

jjbayer committed:
wip: test_query_builder passes

jjbayer committed:
fix: test_organization_metrics

jjbayer committed:
feat: Add syntactic sugar for ops

jjbayer committed:
ref: QueryDefinition

jjbayer committed:
wip

jjbayer committed:
Merge remote-tracking branch 'origin/master' into feat/metrics-interface

jjbayer committed:
ref: MetricsQuery -> QueryDefinition

jjbayer committed:
Merge remote-tracking branch 'origin/master' into feat/metrics-interface

jjbayer committed:
ref: Remove unused type

jjbayer committed:
fix: get_intervals

jjbayer committed:
ref: Remove QueryType flag

jjbayer committed:
ref: Remove helper classes

jjbayer committed:
Merge remote-tracking branch 'origin/master' into feat/metrics-interface

jjbayer committed:
ref: Base sessions v2 API on metrics query builder

jjbayer committed:
ref: cleanup

jjbayer committed:
Fix: Use same orderby & limit as sessions impl

jjbayer committed:
fix: Limit type

jjbayer committed:
Merge branch 'master' into feat/metrics-interface

jjbayer committed:
ref: Make mypy happy

jjbayer committed:
Merge remote-tracking branch 'origin/feat/metrics-interface' into feat/metrics-sessions-v2-wrapper

jjbayer committed:
Merge remote-tracking branch 'origin/master' into feat/metrics-sessions-v2-wrapper

jjbayer commented on src/sentry/snuba/sessions_v2.py:
```diff
@@ -283,7 +283,7 @@ def __init__(self, query, params, allowed_resolution: AllowedResolution):
         for i, field in enumerate(self.fields.values()):
             columns = field.get_snuba_columns(raw_groupby)
-            if i == 0:
+            if i == 0 or field == "sum(session)":  # Prefer first, but sum(session) always wins
```
This is needed to keep consistency between how the two implementations (sessions and metrics) sort the results (see test case below). The previous metrics-based implementation simply ran another snuba query when the first one returned an empty result, but we cannot do that with the current implementation because we now only make a single `MetricsQuery`. https://github.com/getsentry/sentry/blob/7e1e3a904c0cc3a2b0ace7e7df7eb8bc9b4c7498/tests/snuba/api/endpoints/test_organization_sessions.py#L882-L885.

jjbayer committed:
fix: test

untitaker reviewed (approved):
i did a quick scan and it roughly makes sense. i would say we just make sure that the release health comparison doesn't spike and we should be good. we have decent test coverage overall

jjbayer commented on src/sentry/release_health/metrics_sessions_v2.py:
```diff
@@ -705,138 +289,85 @@ def run_sessions_query(
+        query.start,
+        query.end,
+        Granularity(query.rollup),
+        where=_get_filter_conditions(query.conditions),
```
Agreed, what we should expose is a limited condition language for metrics tags.. but that's a story for a separate PR.

jjbayer committed:
fix: Generate default groups

jjbayer committed:
Merge remote-tracking branch 'origin/master' into feat/metrics-sessions-v2-wrapper

jjbayer committed:
feat(sessions): sessions v2 order by (#33857)

Give the sessions API the ability to sort the returned groups by one of
the fields, e.g.:

?field=sum(session)&groupBy=release&orderBy=-sum(session) # descending

Caveats:

You can only order by a single field (e.g. sum(session)), not by group
by values such as release or environment. This only works with the
MetricsReleaseHealthBackend configured. orderBy only works when
session.status is not in groupBy.
